File: wood_factory/wood_factory/page/factory_performance/factory_performance.js
frappe.pages["factory-performance"].on_page_load = function (wrapper) {
    const page = frappe.ui.make_app_page({parent: wrapper, title: __("Factory Performance"), single_column: true});
    page.main.addClass("factory-performance-page");
    page.add_field({label: __("From Date"), fieldname: "from_date", fieldtype: "Date", default: frappe.datetime.add_days(frappe.datetime.get_today(), -30), change: () => load_performance(page)});
    page.add_field({label: __("To Date"), fieldname: "to_date", fieldtype: "Date", default: frappe.datetime.get_today(), change: () => load_performance(page)});
    page.set_primary_action(__("Refresh"), () => load_performance(page));
    load_performance(page);
};

function load_performance(page) {
    frappe.call({method: "wood_factory.wood_factory.page.factory_performance.factory_performance.get_performance_data", args: {from_date: page.fields_dict.from_date.get_value(), to_date: page.fields_dict.to_date.get_value()}, freeze: true}).then(r => render_performance(page, r.message || {}));
}

function render_performance(page, data) {
    const s = data.summary || {};
    page.main.html(`<div class="performance-kpis">${kpi(__("Orders Measured"), s.completed_orders)}${kpi(__("Completed Stages"), s.completed_stages)}${kpi(__("Actual Work Hours"), s.work_hours)}${kpi(__("Blocked Hours"), s.blocked_hours)}</div><section><h3>${__("Stage Performance")}</h3><div class="performance-table">${stage_rows(data.stages || [])}</div></section><section><h3>${__("Likely Bottlenecks")}</h3><div class="bottleneck-grid">${bottleneck_rows(data.bottlenecks || [])}</div></section><section><h3>${__("Worker Activity")}</h3><div class="performance-table">${worker_rows(data.workers || [])}</div></section>`);
}

function kpi(label, value) { return `<div class="performance-kpi"><span>${label}</span><b>${value || 0}</b></div>`; }
function stage_rows(rows) { return `<div class="performance-row header"><b>${__("Stage")}</b><b>${__("Completed")}</b><b>${__("Avg. Minutes")}</b><b>${__("Blocked %")}</b></div>` + rows.map(row => `<div class="performance-row"><span>${__(row.stage)}</span><span>${row.completed}</span><span>${row.avg_minutes}</span><span>${row.blocked_percent}%</span></div>`).join(""); }
function bottleneck_rows(rows) { return rows.length ? rows.map((row, index) => `<div class="bottleneck-card"><small>#${index + 1}</small><h4>${__(row.stage)}</h4><b>${row.avg_minutes} ${__("min avg")}</b><span>${row.blocked_percent}% ${__("blocked")}</span></div>`).join("") : `<div class="performance-empty">${__("Not enough completed production data yet")}</div>`; }
function worker_rows(rows) { return `<div class="performance-row header"><b>${__("Worker")}</b><b>${__("Completed Stages")}</b><b>${__("Work Hours")}</b><b>${__("Blocked Hours")}</b></div>` + rows.map(row => `<div class="performance-row"><span>${frappe.utils.escape_html(row.user)}</span><span>${row.completed_stages}</span><span>${(row.work_minutes / 60).toFixed(2)}</span><span>${(row.blocked_minutes / 60).toFixed(2)}</span></div>`).join(""); }

File: wood_factory/wood_factory/page/factory_performance/factory_performance.py
import frappe
from frappe.utils import add_days, flt, getdate, nowdate


STAGES = ["Cutting", "Edge Banding", "Drilling", "Assembly", "Quality Inspection", "Packing"]


@frappe.whitelist()
def get_performance_data(from_date=None, to_date=None):
    to_date = getdate(to_date or nowdate())
    from_date = getdate(from_date or add_days(to_date, -30))
    if from_date > to_date:
        frappe.throw("From Date cannot be after To Date")

    rows = frappe.db.sql("""
        select s.stage, s.actual_minutes, s.blocked_minutes, s.responsible,
               s.started_at, s.completed_at, s.parent as factory_order
        from `tabFactory Order Stage` s
        where s.status = 'Completed'
          and s.completed_at between %(from_date)s and date_add(%(to_date)s, interval 1 day)
        order by s.completed_at asc
    """, {"from_date": from_date, "to_date": to_date}, as_dict=True)

    stage_map = {stage: {"stage": stage, "completed": 0, "work_minutes": 0, "blocked_minutes": 0, "avg_minutes": 0, "blocked_percent": 0} for stage in STAGES}
    user_map = {}
    total_work = total_blocked = 0
    orders = set()

    for row in rows:
        metric = stage_map.setdefault(row.stage, {"stage": row.stage, "completed": 0, "work_minutes": 0, "blocked_minutes": 0, "avg_minutes": 0, "blocked_percent": 0})
        work = flt(row.actual_minutes); blocked = flt(row.blocked_minutes)
        metric["completed"] += 1; metric["work_minutes"] += work; metric["blocked_minutes"] += blocked
        total_work += work; total_blocked += blocked; orders.add(row.factory_order)
        user = row.responsible or "Unassigned"
        worker = user_map.setdefault(user, {"user": user, "completed_stages": 0, "work_minutes": 0, "blocked_minutes": 0})
        worker["completed_stages"] += 1; worker["work_minutes"] += work; worker["blocked_minutes"] += blocked

    for metric in stage_map.values():
        metric["work_minutes"] = flt(metric["work_minutes"], 2); metric["blocked_minutes"] = flt(metric["blocked_minutes"], 2)
        metric["avg_minutes"] = flt(metric["work_minutes"] / metric["completed"], 2) if metric["completed"] else 0
        elapsed = metric["work_minutes"] + metric["blocked_minutes"]
        metric["blocked_percent"] = flt(metric["blocked_minutes"] / elapsed * 100, 2) if elapsed else 0

    workers = sorted(user_map.values(), key=lambda row: (-row["completed_stages"], row["work_minutes"]))
    bottlenecks = sorted(stage_map.values(), key=lambda row: (-row["avg_minutes"], -row["blocked_percent"]))
    return {
        "period": {"from_date": from_date, "to_date": to_date},
        "summary": {"completed_orders": len(orders), "completed_stages": len(rows), "work_hours": flt(total_work / 60, 2), "blocked_hours": flt(total_blocked / 60, 2)},
        "stages": list(stage_map.values()),
        "workers": workers[:20],
        "bottlenecks": [row for row in bottlenecks if row["completed"]][:3],
    }
